flare/algorithm_zoo/ddpg.py

```diff
@@ -0,0 +1,168 @@
+from copy import deepcopy
+import math
+from flare.framework.algorithm import Algorithm, Model
+from flare.framework import common_functions as comf
+import torch
+import torch.nn as nn
+import torch.optim as optim
+
+
+def fanin_init(m):
+    fan_in, _ = nn.init._calculate_fan_in_and_fan_out(m.weight)
+    bound = 1.0 / math.sqrt(fan_in)
+    nn.init.uniform_(m.weight, -bound, bound)
+    nn.init.uniform_(m.bias, -bound, bound)
+
+
+def uniform_init(m, bound):
+    nn.init.uniform_(m.weight, -bound, bound)
+    nn.init.uniform_(m.bias, -bound, bound)
+
+
+class ContinuousDeterministic(Model):
+    """
+    Model used by DDPG.
+    """
+
+    def __init__(self, input_dims, action_dims):
+        super(ContinuousDeterministic, self).__init__()
+        self.input_dims = input_dims
+        self.action_dims = action_dims
+
+        self.policy_fc1 = nn.Linear(input_dims, 400)
+        self.policy_fc2 = nn.Linear(400, 300)
+        self.policy_fc3 = nn.Linear(300, action_dims)
+        fanin_init(self.policy_fc1)
+        fanin_init(self.policy_fc2)
+        uniform_init(self.policy_fc3, 3e-3)
+
+        self.critic_fc1 = nn.Linear(input_dims, 400)
+        self.critic_fc2 = nn.Linear(400 + action_dims, 300)
+        self.critic_fc3 = nn.Linear(300, 1)
+        fanin_init(self.critic_fc1)
+        fanin_init(self.critic_fc2)
+        uniform_init(self.critic_fc3, 3e-4)
+
+        self.relu = nn.ReLU()
+
+    def get_input_specs(self):
+        return [("sensor", dict(shape=[self.input_dims]))]
+
+    def get_action_specs(self):
+        return [("action", dict(shape=[self.action_dims]))]
+
+    def policy(self, inputs, states):
+        out = self.policy_fc1(inputs.values()[0])
+        out = self.relu(out)
+        out = self.policy_fc2(out)
+        out = self.relu(out)
+        action = self.policy_fc3(out)
+        return dict(action=action), states
+
+    def value(self, inputs, states, actions=None):
+        out = self.critic_fc1(inputs.values()[0])
+        h = self.relu(out)
+        if actions is None:
+            actions, _ = self.policy(inputs, states)
+        a = actions["action"]
+        out = self.critic_fc2(torch.cat([h, a], dim=-1))
+        out = self.relu(out)
+        value = self.critic_fc3(out)
+        return dict(q_value=value), states
+
+    def policy_optimizer(self, optim_specs):
+        return optim_specs[0]([{
+            'params': self.policy_fc1.parameters()
+        }, {
+            'params': self.policy_fc2.parameters()
+        }, {
+            'params': self.policy_fc3.parameters()
+        }], **optim_specs[1])
+
+    def critic_optimizer(self, optim_specs):
```

Comment:
Comments. Why having this weird behavior?

Reply:
Not addressed? Why functions policy_optimizer() and critic_optimizer() are implemented in a weird way.